java/dagger/android/proguardprocessor/ProguardProcessor.java

```diff
@@ -0,0 +1,117 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package dagger.android.proguardprocessor;
+
+import com.google.auto.service.AutoService;
+import java.io.IOException;
+import java.io.Writer;
+import java.util.Set;
+import javax.annotation.processing.AbstractProcessor;
+import javax.annotation.processing.Filer;
+import javax.annotation.processing.Processor;
+import javax.annotation.processing.RoundEnvironment;
+import javax.annotation.processing.SupportedAnnotationTypes;
+import javax.lang.model.SourceVersion;
+import javax.lang.model.element.Element;
+import javax.lang.model.element.TypeElement;
+
+import static javax.tools.StandardLocation.CLASS_OUTPUT;
+
+/**
+ * An {@linkplain Processor annotation processor} to generate dagger-android's specific proguard
+ * needs. This is only intended to run over the dagger-android project itself, as the alternative
+ * is to create an intermediary java_library for proguard rules to be consumed by the project.
+ *
+ * <p>Basic structure looks like this:
+ * <pre><code>
+ *   resources/META-INF/com.android.tools/proguard/dagger-android.pro
+ *   resources/META-INF/com.android.tools/r8/dagger-android.pro
+ *   resources/META-INF/proguard/dagger-android.pro
+ * </code></pre>
+ */
+@SupportedAnnotationTypes(ProguardProcessor.PLACEHOLDER_ANNOTATION)
+@AutoService(Processor.class)
+public final class ProguardProcessor extends AbstractProcessor {
+
+  static final String PLACEHOLDER_ANNOTATION = "dagger.android.internal.ProguardRulesReceiver";
+
+  @Override
+  public boolean process(Set<? extends TypeElement> annotations, RoundEnvironment roundEnv) {
+    if (roundEnv.errorRaised()) {
+      return false;
+    } else if (annotations.isEmpty()) {
+      return false;
+    }
+
+    Set<? extends Element> elements =
+        roundEnv.getElementsAnnotatedWith(processingEnv.getElementUtils()
+            .getTypeElement(PLACEHOLDER_ANNOTATION));
+    if (elements.isEmpty()) {
+      return false;
+    } else {
+      generate();
+      return false;
+    }
+  }
+
+  private void generate() {
+    Filer filer = processingEnv.getFiler();
+
+    StringBuilder rulesBuilder = new StringBuilder();
+    rulesBuilder.append("-dontwarn com.google.errorprone.annotations.**")
+        .append("\n");
+
+    String proguardRules = rulesBuilder.toString();
+
+    String r8Rules = rulesBuilder.append(
+        "-identifiernamestring class dagger.android.internal.AndroidInjectionKeys {")
+        .append("\n")
+        .append("  java.lang.String of(java.lang.String);")
+        .append("\n")
+        .append("}")
+        .append("\n")
+        .toString();
+
+    try {
+      // Write META-INF/com.android.tools/proguard/dagger-android.pro
+      try (Writer writer = filer.createResource(CLASS_OUTPUT,
+          "",
+          "META-INF/com.android.tools/proguard/dagger-android.pro")
+          .openWriter()) {
+        writer.write(proguardRules);
+      }
```

Comment:
Extract to a method that takes the file path and content? So that we don't repeat the same writing maneuver.

Reply:
2e43102